common/core/hwclayer.cpp

```diff
@@ -118,9 +118,7 @@ void HwcLayer::SetSurfaceDamage(const HwcRegion& surface_damage) {
       return;
     }
   } else if (rects == 0) {
-    rect = display_frame_;
-    // damage is assigned with display_frame, no need to transform
-    state_ &= ~kSurfaceDamageChanged;
+    rect = source_crop_;
```

Comment:
Does this WA actually disable partial update?

Reply:
No, partial update are only disabled when the left and top of source_crop are not 0.
Partial-refresh still work when the surface-damage is set and source_crop.left and top are 0.

Previously, HWC suppose SF send the 3 attribute, surface_damage, source_crop, and display_frame for each layer in each frame.
From Q, it is not true. SF send a layer only with source_crop and display_frame. surface_damage could not be set by SF.
So use the flag kSurfaceDamageChanged to mark if the surface_damage is set or not, if it is not set, assigned with display_frame. 

Once it is set with a non-empty array, use this surface-damage if source_crop.top and left are 0 (Most case, the source_crop are 0,0, X, X). if source_crop.top and left are not 0, use WA

Once it is set with a empty array, it mean the whole layer need to be updated. assigned with source_crop here for transforming